openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateZonePropertyUserViewFactorsBySurfaceName.cpp

```diff
@@ -0,0 +1,140 @@
+
+namespace openstudio {
+
+namespace energyplus {
+
+OptionalModelObject ReverseTranslator::translateZonePropertyUserViewFactorsBySurfaceName( const WorkspaceObject & workspaceObject )
+{
+  if( workspaceObject.iddObject().type() != IddObjectType::ZoneProperty_UserViewFactors_bySurfaceName ){
+    LOG(Error, "WorkspaceObject is not IddObjectType: ZonePropertyUserViewFactorsBySurfaceName");
+    return boost::none;
+  }
+
+  boost::optional<WorkspaceObject> target = workspaceObject.getTarget(ZoneProperty_UserViewFactors_bySurfaceNameFields::ZoneorZoneListName);
+
+  std::vector<ThermalZone> thermalZones;
+
+  if (target){
+
+    boost::optional<ModelObject> mo;
+    if (target->iddObject().type() == IddObjectType::Zone){
+      mo = translateAndMapWorkspaceObject(target.get());
+      if( mo ) {
+        if( boost::optional<Space> space = mo->optionalCast<Space>() ) {
+          boost::optional<ThermalZone> thermalZone = space->thermalZone();
+          if (thermalZone){
+            thermalZones.push_back(*thermalZone);
+          }
+        }
+      }
+    }
+  }
+
+  if(thermalZones.empty())
+  {
+    LOG(Error, "Error importing object: "
+             << workspaceObject.briefDescription()
+             << " Can't find associated ThermalZone.");
+
+    return boost::none;
+  }
+
+  boost::optional<ModelObject> result;
+  for (ThermalZone thermalZone : thermalZones){
+
+    openstudio::model::ZonePropertyUserViewFactorsBySurfaceName zoneProp = thermalZone.getZonePropertyUserViewFactorsBySurfaceName();
+
+    if (!result){
+      result = zoneProp;
+    }
+
+    for (const IdfExtensibleGroup& idfGroup : workspaceObject.extensibleGroups()){
+      WorkspaceExtensibleGroup workspaceGroup = idfGroup.cast<WorkspaceExtensibleGroup>();
+
+      boost::optional<WorkspaceObject> fromTarget = workspaceGroup.getTarget(ZoneProperty_UserViewFactors_bySurfaceNameExtensibleFields::FromSurface);
+      boost::optional<WorkspaceObject> toTarget = workspaceGroup.getTarget(ZoneProperty_UserViewFactors_bySurfaceNameExtensibleFields::ToSurface);
+
+      boost::optional<ModelObject> toModelObject;
+      boost::optional<ModelObject> fromModelObject = translateAndMapWorkspaceObject(*fromTarget);
+      if (fromTarget != toTarget){
+        toModelObject = translateAndMapWorkspaceObject(*toTarget);
+      } else {
+        toModelObject = fromModelObject;
+      }
+
+      OptionalDouble _viewFactor = workspaceGroup.getDouble(ZoneProperty_UserViewFactors_bySurfaceNameExtensibleFields::ViewFactor);
+
+      // If the objects translated ok, and the viewFactor double value exists
+      if (fromModelObject && toModelObject && _viewFactor) {
+        // try to create a viewFactor, which will throw in case wrong type, or wrong double value. Here we let it slide though
+        try {
+          ViewFactor viewFactor(fromModelObject.get(), toModelObject.get(), _viewFactor.get());
+          // add the view factor
+          if (zoneProp.addViewFactor(viewFactor)) {
```

Comment:
Did you try out the RT in OSApp? Should this be `!zoneProp.addViewFactor(viewFactor)`? Seems like this warning is thrown if the viewFactor is added.

Reply:
Yes please change that. Good catch